### PR TITLE
PrePublishPanel: suggest tags and postformat

### PR DESCRIPTION
**Pull request by oandregal**
Related https://github.com/WordPress/gutenberg/issues/7426

This PR aims to include the Tags and the PostFormat components in the pre-publish panel:

* Tags component is shown only if the user hasn't tagged the post yet.
* PostFormat is shown if the suggested PostFormat is different from the current one.

### How to test

* Create a new post.
* Add a block image (doesn't need to have any actual image).
* Click Publish. 
* The expected behavior is that the pre-publish panel includes both the Tags and the PostFormat components to prompt the user to add them. Both should be collapsed. See:

| Closed | Opened |
| --- | --- |
| pre-publish-panel-copy-long-closed | pre-publish-panel-copy-toffumatt |

Test variations of the above: with the document sidebar closed (as this is when tags are fetched), with tags and `image` as the current post format, etc. The suggested post format is shown below the current post format, if there is any suggestion - typing `wp.data.select('core/editor').getSuggestedPostFormat()` in the console will give you the same result (or `null` if none is to be suggested).

### TODO

- [x] Make text translatable.
- [x] Try a more conversational post format suggestion (see https://github.com/WordPress/gutenberg/issues/7426#issuecomment-403977765)
- [x] Not shown conditions:
  - [x] Post already has some tag applied.
  - [x] It's a page, not a post (more generally, if the post type allows tags).
- [x] Fix: 
  - [x] Don't crash if tags weren't fetched yet.
  - [x] Panel is hidden after adding one tag in the pre-publish panel.

### Open questions 

- [x] Should we suggest only tags or any available taxonomy? We're going to go with tags as a starter.

### CODE REVIEW

**Comment by nosolosw** [comment]
To the open question regarding whether to include only tags or all available taxonomies in the PrePublishPanel: I lean towards adding any taxonomy that has at least one term (is there any other heuristic we can use to infer which taxonomies are relevant for a given site?).

**Comment by chrisvanpatten** [comment]
> To the open question regarding whether to include only tags or all available taxonomies in the PrePublishPanel: I lean towards adding any taxonomy that has at least one term (is there any other heuristic we can use to infer which taxonomies are relevant for a given site?)

I would just make it tags or categories, but nothing else by default. Perhaps make it filterable so developers can opt their taxonomies in, but I think it's dangerous to assume all other taxonomies are relevant or worthy of a notification.

**Comment by sarahmonster** [comment]
I think the original concept here (cc @mtias) was to show the tags module only when no tags have been added to the post—so it wouldn't appear at all times, but it's there to serve as a reminder if the user has forgotten to add any tags. 

I’m not sure categories are as relevant to the modern #hashtag web—and I suspect people are often unsure of the difference between the two—so I think starting with tags would be sensible. (I see a *lot* of people who don't even use categories, but I'm not sure if that's because the categories are less surfaced in the UI or because they're just an advanced feature.)

The *ideal* would probably be if we suggested some tags based on the post content and/or the users’ commonly used tags, but that’s obviously a bit more complicated to implement. Failing that, suggesting the most commonly-used tags (maybe three?) would be a good start. 

I’m not certain if tag suggestions are better in the pre-publish panel or in the post-publish panel, since they aren't really critical to the pre-publish experience, per se. Before publishing, I want to make sure the date is right, check on anything that's going to automatically send out... basically make sure I have control over everything that happens "immediately." Adding tags can help users find your post, but ultimately it doesn't matter much if you add them before or after publishing the post. 

That said, users are likely to be a bit more focussed at the pre-publish stage, and since we currently aren't bombarding them with information and Things to Check at this stage, having them here for the time being probably works. If this panel ends up being used heavily by plugins, we may want to consider shifting them to the post-publish panel in order to reduce noise and make publishing as frictionless as possible.

**Comment by nosolosw** [comment]
@sarahmonster this is ready for design review.

cc @gziolo for code review.

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (gziolo) [inline packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: Can you help me understand why it is not possible to set state only once on initial load and keep it this way for the rest of interaction with this panel? Checks like:
```js
if ( currentHadTags === undefined && hasTags !== undefined ) {
```
indicate that it can be `undefined` for some reasons. If that is the reason, can we postpone mounting this component until this data is fetched?

Comment (nosolosw) [reply packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: That's explained it in the long `getDerivedStateFromProps` comment. TLDR is: tags may be not fetched yet at this point. How to test this behavior:

- Create a new post.
- Make sure the sidebar document hasn't been opened. If it is, collapse it, and reload the page.
- Click the publish button. At this point, the pre-publish panel will be opened but the taxonomies aren't fetched yet (because the sidebar document is closed).

I'd love to get rid of this complexity, but I'm a bit fuzzy about Gutenberg fetching mechanisms. Any suggestion about code examples I should look at for inspiration and learn-up?

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (youknowriad) [inline packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: Can't this condition be something like that instead:

```
const tagTaxonomy = select( 'core' ).getTaxonomy( 'post_tag' );
const shouldShowMoreTags = tagTaxonomy && ( ! select( 'core/editor' ).getEditedPostAttribute( tags.rest_base ) || select( 'core/editor' ).getEditedPostAttribute( tags.rest_base ).length ===0 ) &&  tagTaxonomy.types.some( ( type ) => type === postType );
```

Something in that vein?

And I guess we can leverage the `ifCondition` HigherOrderComponent to avoid the `MaybeTagsPanel` component entirely?

**Comment by nosolosw** [comment]
I was thinking of adding the PostFormat component in a different PR but figured out it'd better to include it here as well. See cd57766. It helps to review the changes holistically (things like whether the panels should be opened/close, see the space they take, and so on). I also hope this minimizes code review time.

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (nosolosw) [reply packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: Found a way to do this by using the `ifCondition` HOC component suggested by Riad, see e7dde0f.

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (nosolosw) [reply packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: Thanks for the `ifCondition` suggestion: I've used that in e66d09d to prevent the component from mounting if some conditions aren't met.

In this case, the internal component state is useful: we need a place to encode whether the post had tags *when the user clicked publish*. We can't use the `hasTags` prop because it'll become `true` if the user adds a new tag in the pre-publish panel, causing a re-render that hides the tags component, and so preventing the user from adding more than one tag.

I've also refactored the logic to make the assumptions more explicit.

File: packages/editor/src/components/post-publish-panel/prepublish.js
Comment (gziolo) [inline packages/editor/src/components/post-publish-panel/prepublish.js]: Have you checked how `<PostSchedule /` or `<PostVisibility />` are implemented?

https://github.com/WordPress/gutenberg/tree/master/packages/editor/src/components/post-schedule
https://github.com/WordPress/gutenberg/tree/master/packages/editor/src/components/post-visibility

They contain two parts the components which render the panel's body and a wrapper component with suffix `Check`. It would be nice to align.

@youknowriad, why we don't wrap those components with checks in the publish panel? We do it in the sidebar:
https://github.com/WordPress/gutenberg/blob/master/edit-post/components/sidebar/post-schedule/index.js#L15-L35

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (youknowriad) [reply packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: Oh ok, I get it now and in that case, I think you were right originally and now it can fail if the taxonomy is not loaded initially (in theory, because in practice, it's already there).

But I'm wondering if the `hasTags` condition makes sense at all, why can't we just always show the tags there if the post supports it? It seems better for me even if I already entered tags to check them on pre-publish. It also makes the code way simpler.

File: packages/editor/src/components/post-publish-panel/prepublish.js
Comment (youknowriad) [reply packages/editor/src/components/post-publish-panel/prepublish.js]: Because it's already taken care of in the component itself. In the sidebar, we don't want the extra panels if the check is false, here, there's no extra wrappers added.

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (nosolosw) [reply packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: > Oh ok, I get it now and in that case, I think you were right originally and now it can fail if the taxonomy is not loaded initially (in theory, because in practice, it's already there).

It's still correct. Note that if the taxonomy is not loaded (which do happen when the user clicks the publish button without the sidebar being opened first), the `areTagsFetched` will be false. https://github.com/WordPress/gutenberg/commit/e66d09dc547fed395e7e9182250f9c451aaf33f4 decoupled two separate concerns that were previously encoded in only one variable: whether or not the tags taxonomy has been fetched (`areTagsFetched`) and if the post has tags (`hasTags`).

> It seems better for me even if I already entered tags to check them on pre-publish. It also makes the code way simpler.

That's a design decision that I defer to Matías and Sarah. It seems sensible to limit the amount of info in the pre-publish panel - we don't want to have here info that's already filled in the sidebar by the user.

File: packages/editor/src/components/post-publish-panel/prepublish.js
Comment (nosolosw) [reply packages/editor/src/components/post-publish-panel/prepublish.js]: I can extract the `ifCondition` to a `Check` component if that's preferred. We use a variety of patterns through the codebase, though, so unless we want to consolidate towards using a particular one I'm happy as it is.

**Comment by nosolosw** [comment]
Changes introduced in 46a181a

- gave dashicon a size of 18 (it was 20 previously)
- added the yellow color
- updated copy

| Before | After |
| --- | --- |
| post-format-and-tags | tips-before-after-18 |

**Comment by nosolosw** [comment]
Changes in b4de407 make the post format suggestion more conversational:

| Before | After |
| --- | --- |
| pre-publish-panel-open | pre-publish-panel-conversational |

**Comment by nosolosw** [comment]
7ffe43a and 06f2b7f update the copy text:

| Before | After |
| --- | --- |
| pre-publish-panel-conversational | pre-publish-panel-copy |

**Review by sarahmonster — CHANGES_REQUESTED** [review]
I know I suggested it, but I'm definitely not crazy about the yellow lightbulb. The yellow is way too much, and the icon feels a bit awkward. We could try replicating the tips that are used elsewhere (ie, a blue dot), but I think that may be a bit confusing. Let's just nix the icon altogether for now.

>  Your readers will be able to find your posts more easily. 

Suggestion: 

> This will help readers find your posts.

I also don't seem to be able to get the post format suggestion to display:

```
PostFormatSuggested@http://localhost:8888/wp-content/plugins/gutenberg/build/editor/index.js?ver=1533257225:23894:1
yh@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:95:430
lg@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:120:88
mg@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:120:386
gc@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:127:202
vb@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:126:230
ub@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:126:65
wg@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:137:182
Ze@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:41:24
EventListener.handleEvent*p@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:40:392
W@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:49:377
qf@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:52:256
zh@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:108:355
kg@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:118:263
lg@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:120:121
mg@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:120:386
gc@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:127:202
vb@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:126:230
ub@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:126:65
zd@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:124:449
ra@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:123:319
enqueueSetState@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react-dom.min.82e21c65.js:189:231
q.prototype.setState@http://localhost:8888/wp-content/plugins/gutenberg/vendor/react.min.ab6b06d4.js:18:428
subscribe/this.unsubscribe<@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:7467:13
globalListener/<@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:8027:14
globalListener@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:8026:5
registerReducer/<@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:8068:9
dispatch@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:6065:7
dispatch@http://localhost:8888/wp-admin/post-new.php:1:38397
_callee2$@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:8217:21
tryCatch@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:6338:37
invoke@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:6564:22
defineIteratorMethods/</prototype[method]@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:6390:16
asyncGeneratorStep@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:437:16
_next@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:459:9
run@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:3228:22
notify/<@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:3245:30
flush@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:1938:9
MutationCallback*./node_modules/core-js/library/modules/_microtask.js/module.exports@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:1957:5
./node_modules/core-js/library/modules/es6.promise.js@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:3166:17
__webpack_require__@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:21:12
./node_modules/core-js/library/fn/promise.js@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:1037:1
__webpack_require__@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:21:12
./node_modules/@babel/runtime/core-js/promise.js@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:203:18
__webpack_require__@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:21:12
./node_modules/@babel/runtime/helpers/asyncToGenerator.js@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:433:16
__webpack_require__@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:21:12
./packages/data/build-module/registry.js@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:7828:97
__webpack_require__@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:21:12
./packages/data/build-module/default-registry.js@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:7503:67
__webpack_require__@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:21:12
./packages/data/build-module/index.js@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:7648:75
__webpack_require__@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:21:12
this.wp.data<@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:70:18
@http://localhost:8888/wp-content/plugins/gutenberg/build/data/index.js?ver=1533257225:2:11
```

Side note: this looks pretty busy right now, but I think could be improved a great deal by merging #7879, which seems to have gotten stuck in limbo. I'll look into pushing that change through in a more incremental way.

**Comment by nosolosw** [comment]
Changes in b379a78, 0fcc49e, and 1f20a64

| Before | After | 
| --- | --- |
| pre-publish-panel-copy-translations | pre-publish-panel-no-icon |

**Comment by michelleweber** [comment]
On the tags, I'd love to add just a few more words that add detail on the what/why of tags:

> Enter a few words that describe your post's subject to help interested readers find it.

And little tweaks on post format:

> Your theme uses post formats -- styling tweaks that complement different kinds of content. The Image format would be great for this post.

**Comment by nosolosw** [comment]
Implemented Michelle's suggestions at c29f3b0

pre-publish-panel-copy-text

**Comment by nosolosw** [comment]
@sarahmonster @gziolo copy, design, and code feedback has been addressed. Shall we land this?

**Review by tofumatt — CHANGES_REQUESTED** [review]
There are some readability and styling issues here (sorry eslint didn't catch them), but more importantly there are gettext issues that will result in poorly-translated strings. They need to be fixed before we merge this.

I had a few questions as well 😄

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (tofumatt) [reply packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: Actually, I get this one is okay, but the above one (`{ __( 'The ' ) } + PostFormatSuggested`) needs fixing...

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (tofumatt) [inline packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: Also: where/what readers will this help? Do we mean on WordPress.com? I think this lacks context and I wouldn't know what it would mean when presented with it.

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (nosolosw) [reply packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: No, it can't be. We need to encode whether the post had tags _when the user clicked publish_ and the pre-publish panel was first shown. The `hasTags` can't be used for that because it will change to `true` if the user adds a new tag within the pre-publish panel. That will cause a re-render of this component and a new `hasTags` check: as a result the panel will be hidden preventing the user from adding more than one tag.

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (nosolosw) [reply packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: :heart: It seems I've got every parenthesis wrong :sweat_smile:

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (nosolosw) [reply packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: I'm ok with being explicit.

File: packages/editor/src/components/post-publish-panel/maybe-post-format-panel.js
Comment (nosolosw) [reply packages/editor/src/components/post-publish-panel/maybe-post-format-panel.js]: Thanks for the insight! I was actually unsure about this. I showed [my concerns](https://github.com/WordPress/gutenberg/pull/8235#issuecomment-409997482) before the copy text feedback and assumed that if the dinamyc text was placed at the beginning or end that'd be fine. Good catch :star2:

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (nosolosw) [reply packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: e150fbc

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (nosolosw) [reply packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: e150fbc

File: packages/editor/src/components/post-publish-panel/maybe-post-format-panel.js
Comment (nosolosw) [reply packages/editor/src/components/post-publish-panel/maybe-post-format-panel.js]: e8f4408

**Comment by nosolosw** [comment]
@sarahmonster @michelleweber we needed to adjust a bit the copy text as per translation concerns. TLDR: links can't be injected in the middle of the sentence at this moment, so the whole sentence will have to be a link. For this reason, it seems more appropriate to use a copy that states that it is an action. Does this copy sounds OK to you?

pre-publish-panel-copy-translations

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (nosolosw) [reply packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: I'm OK with this but I'll defer to what Sarah/Michelle think here: https://github.com/WordPress/gutenberg/pull/8235#issuecomment-411387138

**Comment by michelleweber** [comment]
Yup, works for me!

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (mtias) [reply packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: Could `prevProps` help there to keep it visible?

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (tofumatt) [reply packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: Fair enough, I guess when it's the render method the explicitness helps too. I thought of this as sort of an early return but I guess it's less about control flow and more about content rendering, so good point 👍

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (nosolosw) [reply packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: `prevProps` would be enough up to the second tag, at that point, we'll have the same problem: `prevProps.hasTags` will be `true`.

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (sarahmonster) [reply packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: This isn't for WordPress.com at all. Tags are good for SEO and to help users find related posts across your own site. [This](https://yoast.com/tagging-posts-properly-for-users-and-seo/) is a pretty good article if we want to add more context, but I'm wary about linking to anything from a UI that isn't coming directly from WordPress.

Some copyedit suggestions for this and other parts of this UI below, in the interests of keeping everything in one place.

**Comment by sarahmonster** [comment]
Some thoughts on copy:

I'm not sure we should use "Tip" here, because "tip" already means something different in the UI (overlaid tips that are part of the NUX). That said, I like having some sort of delineation to indicate that the panel won't always be there—that it's there to provide ideas/suggestions/guidance.  What about if we use "Suggestion:" here instead, and shorten the actual suggestion text? So "Suggestion: Add tags" and "Suggestion: Use a post format".

For the tags, we say:

> Enter a few words that describe your post’s subject. This will help interested readers find it.

Let's give a bit more context here, so people understand why we're suggesting this:
> Tags help users and search engines navigate your site and find your content. Add a few keywords to describe your post.

For the post format, we say:

> Your theme uses Post Formats, styling tweaks that highlight different kinds of content like images or videos.

I think here, we probably need even more context, because post formats are weird, and, while most people know what a tag is (thanks Instagram!), post formats are probably less easily grokked.

What about something like:
> Your theme uses post formats to highlight different kinds of content, like images or videos. Apply a post format to see this special styling.

> Apply the [whatever] format.

Note that this PR will look a whole lot better with #7879 merged, because right now the lack of any separator between the panels makes it look really messy. Also, it's time for "Here, you can do a last-minute check up of your settings below, before you publish." to wander off into the sunset. 🌅

**Comment by nosolosw** [comment]
This is how it'd look like the changes proposed by @sarahmonster  (31ce0d5)

pre-publish-panel-long-copy

File: packages/editor/src/components/post-publish-panel/maybe-post-format-panel.js
Comment (tofumatt) [inline packages/editor/src/components/post-publish-panel/maybe-post-format-panel.js]: This isn't exactly as suggested by @sarahmonster (it's lacking a period and a comma, as well as using "one" instead of the clearer "a post format"). I've pushed a tweak though 😄

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (tofumatt) [reply packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: In any project that uses Redux I'm always sad about using component `state` because we lose out on Redux debugging features. I also think it's nicer when things are all props over using component state, but maybe that's just me...

I can see this being more complex if you dispatched an action based on whether this post had any tags when the preview panel appeared, but that's instinctively what I'd prefer.

-----

Anyway, failing that, I think this could be better-named. Because it's similarly named to `hasTags` I thought it was just an assignment to `state` and was confused. Maybe:

```js
class MaybeTagsPanel extends Component {
	constructor( props ) {
		super( props );
		this.state = {
			showAddTagsPanel: ! props.hasTags,
		};
	}

	render() {
		if ( this.state.showAddTagsPanel ) {
			return <TagsPanel />;
		}

		// This post had tags before this component was loaded, so
		// we won't show a tag suggestion panel.
		return null;
	}
}
```

How about that?

**Review by tofumatt — CHANGES_REQUESTED** [review]
Okay, I get how and why you're using `state` in the component now. I have an idea for how to clarify the intention of the code a bit. As mentioned I'd prefer us to use actions but failing that I think we can at least make it a bit clearer 😄

File: packages/editor/src/components/post-publish-panel/maybe-post-format-panel.js
Comment (nosolosw) [reply packages/editor/src/components/post-publish-panel/maybe-post-format-panel.js]: I should have used copy-pasta!

Comment (sarahmonster) [reply packages/editor/src/components/post-publish-panel/maybe-post-format-panel.js]: My persnickety heart appreciates this! 🍝❤️

**Comment by nosolosw** [comment]
> In any project that uses Redux I'm always sad about using component state because we lose out on Redux debugging features. I also think it's nicer when things are all props over using component state, but maybe that's just me...

:100: Cannot agree more with this sentiment.

Still, I believe this is one of the few cases where using internal `state` makes sense, as the code becomes simpler to reason about. At the same time, it's now clear to me that this code fails to communicate its purpose, though. That should be refactored, and I'll take a look at your suggestion tomorrow - my brain is clearly off today!

**Comment by tofumatt** [comment]
Yeah, I getcha. I think the amount of added complexity to use Redux here isn't worth it in a project that already uses internal component state quite heavily. If we were 99% pure Redux I'd be stricter, but I think as-is as long as we debate it and then decide it's worth the trade-off (as it probably is here), we're good 😄

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (nosolosw) [reply packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: Hope this helps 0bcaa57

**Comment by nosolosw** [comment]
@sarahmonster @tofumatt can we :ship: this or is there anything else we should address?

**Review by tofumatt — APPROVED** [review]
Thanks for the changes and sorry for all the back-and-forth on this. I dig it! 👍

**Review by sarahmonster — APPROVED** [review]

**Comment by sarahmonster** [comment]
Thanks for all your patience with this, @nosolosw! This looks good by me now. 🚢 !

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (nosolosw) [inline packages/editor/src/components/post-publish-panel/maybe-tags-panel.js]: I learned that pages don't have taxonomies the hard way https://github.com/WordPress/gutenberg/pull/9082

File: packages/editor/src/components/post-publish-panel/maybe-post-format-panel.js
Comment (afercia) [inline packages/editor/src/components/post-publish-panel/maybe-post-format-panel.js]: This misses a translators comment. Will try to address it in #9363

**Pull request merged**
